url969d1ddca4_compmstr_java_scheme_tgz-pJ8-com_undi_javascheme_JavaSchemeJ8/src/com/undi/javascheme/SchemeEval.java

```diff
@@ -0,0 +1,804 @@
+        //Check if varName is a variable
+        try{
+          SchemeObject javaObj = lookupVariableValue(SchemeObject.makeSymbol(varName), originalEnv);
+          return SchemeObject.makeJavaObj(Reflector.getInstanceField(javaObj.getJavaObj(), fieldName));
+        }catch(IllegalArgumentException ex){
+          //Not a variable, try for static field
+          try{
+            return SchemeObject.makeJavaObj(Reflector.getStaticField(varName, fieldName));
+          }catch(Exception e){
+            //Not a field, try for a static method 
+            return SchemeObject.makeJavaStaticMethod(varName, fieldName);
+          }
+        }
+      }
+    }
+    //Check for classes/methods
+    if(var.getSymbol().startsWith(".") && var.getSymbol().length() > 1){
+      System.out.println("Java method/member: " + var.getSymbol().substring(1));
+      return SchemeObject.makeJavaMethod(var.getSymbol().substring(1));
+    }else if(var.getSymbol().endsWith(".")){
+      String className = var.getSymbol().substring(0, var.getSymbol().length() - 1);
+      System.out.println("Java Constructor: " + className);
+      return SchemeObject.makeJavaConstructor(className);
+    }
+    try{
+    	Class cls = Reflector.classFromName(var.getSymbol());
+    	if(cls != null){
+    		return SchemeObject.makeJavaObj(cls);
+    	}
+    }catch(Exception e){ }
+    throw new IllegalArgumentException("Unbound Variable: " + var.getSymbol());
+  }
+  
+  public void setVariableValue(SchemeObject var, SchemeObject val, SchemeObject env){
+    HashMap<SchemeObject, SchemeObject> frameMap;
+    while(!env.isEmptyList()){
+      frameMap = firstFrame(env).getHashMap();
+      if(frameMap.containsKey(var)){
+        frameMap.put(var, val);
+        return;
+      }
+      env = enclosingEnvironment(env);
+    }
+		throw new SchemeException("Unbound Variable - " + var.getSymbol());
+  }
+  
+  public void defineVariable(SchemeObject var, SchemeObject val, SchemeObject env){
+    SchemeObject frame;
+    frame = firstFrame(env);
+    addBindingToFrame(var, val, frame);
+  }
+  
+  public SchemeObject setupEnvironment(){
+    SchemeObject initialEnv;
+    initialEnv = extendEnvironment(SchemeObject.THE_EMPTY_LIST,
+                                      SchemeObject.THE_EMPTY_LIST, 
+                                      emptyEnvironment);
+    
+    return initialEnv;
+  }
+  
+  public SchemeObject eval(SchemeObject exp, SchemeObject env){
+    SchemeObject procedure;
+    SchemeObject arguments;
+    TAILCALL:
+      while(true){
+        if(isSelfEvaluating(exp)){
+          return exp;
+        }else if(isQuoted(exp)){
+          return quoteContents(exp);
+        }else if(isDefinition(exp)){
+          return evalDefinition(exp, env);
+        }else if(isAssignment(exp)){
+          return evalAssignment(exp, env);
+        }else if(isVariable(exp)){
+          try{
+            return lookupVariableValue(exp, env);
+          }catch(IllegalArgumentException ex){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)